ConcurrentAPI.adoc

```diff
@@ -0,0 +1,244 @@
+Acquires the lock if it is available and returns with the value true.
+If the lock is not available then this method with the value false.
+
+###### CompletableFuture<Boolean> tryLock(long time, TimeUnit unit);
+
+Acquires the lock if it is free within the given waiting time and the current thread has not been interrupted.
+If the lock is available this method returns with the value true.
+
+The lock is acquired by the current thread; or
+Some other thread interrupts the current thread, and interruption of lock acquisition is supported; or
+The specified waiting time elapses
+If the lock is acquired then the value true is returned.
+
+Parameters:
+time - the maximum time to wait for the lock
+unit - the time unit of the time argument
+Returns:
+true if the lock was acquired and false if the waiting time elapsed before the lock was acquired
+
+CompletableFuture fails with InfinispanLockException in case of error (InterruptedException, or any other non checked exceptions)
+
+###### CompletableFuture<Boolean> unlock();
+
+If the lock is rentrat, only the Thread holding the lock will be able to unlock, otherwise, any thread can unlock
+and it will behave as a Semaphore with one permit.
+true answer will say that the operation was succesul and the lock has been released, false the lock has not been relased
+
+
+=== Demo (as if the code was blocking)
+
+This main is an example on calling the API as if it was blocking, the corrent implementation should be done
+
+```java
+   public static void main(String[] args) throws Exception {
+        EmbeddedCacheManager cm = Infinispan.createClustered();
+        CounterManager counterManager = EmbeddedCounterManagerFactory.asCounterManager(cm);
+        DistributedLockManager lockManager = EmbeddedLockManagerFactory.asDistributedLockManager(cm);
+        InfinispanLock lock = lockManager.createLock("lock");
+        WeakCounter counter = counterManager.weakCounter("counter");
+
+        for (int i = 0; i < 100; i++) {
+            System.out.println("Counter on " + i + " is => " + counter.getValue());
+            lock.lock().get();
+            try {
+                counter.increment();
+                Thread.sleep(1000);
+                counter.decrement();
+            } finally {
+                lock.unLock();
+            }
+        }
+        cm.stop();
+    }
+
+```
+
+
+
+== (preview) Distributed CountDownLatch
+
+A synchronization aid that allows one or more threads to wait until a set of operations being performed in other threads completes.
+
+
+https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html
+
+
+=== CountDownLatchManager and configuration
+
+```java
+
+public class CountDownLatchManager {
+
+   InfinispanCountDownLatch define(CountDownLatchConfiguration configuration);
+   InfinispanCountDownLatch get(String name);
+   boolean remove(String name);
+
+}
+
+public interface InfinispanCountDownLatch {
+   CompletableFuture<Void> await();
```

Comment:
Either the use of `CompletableFuture` or `await` does not sound right, as this latter evokes blocking and it could be easily misused. Maybe 'asyncWait`?
The second overload won't be needed with Java 9's `orTimeout` but it's useful for now. Just make sure that you don't create `ScheduledThreadPool` per instance :)